src/Mvc/Mvc.Analyzers/src/Startup/ServicesAnalysis.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.AspNetCore.Analyzers
+{
+    internal class ServicesAnalysis : ConfigureServicesMethodAnalysis
+    {
+        public static ServicesAnalysis CreateAndInitialize(OperationBlockStartAnalysisContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var analysis = new ServicesAnalysis((IMethodSymbol)context.OwningSymbol);
+
+            var services = ImmutableArray.CreateBuilder<ServicesItem>();
+            context.RegisterOperationAction(context =>
+            {
+                // We're looking for usage of extension methods, so we need to look at the 'this' parameter
+                // rather than invocation.Instance.
+                if (context.Operation is IInvocationOperation invocation &&
+                    invocation.Instance == null &&
+                    invocation.Arguments.Length >= 1 &&
+                    invocation.Arguments[0].Parameter.Type.Name == "IServiceCollection")
```

Comment:
In terms of error checking, do we need to verify `Parameter`, `Type` are defined? For instance if the analyzer is invoked when the code is incomplete or when symbols are missing, would this null ref?

Reply:
That's a great question.